enterprise/server/build_event_publisher/build_event_publisher.go

```diff
@@ -88,98 +93,133 @@ func (p *Publisher) run(ctx context.Context) {
-						EventTime: ptypes.TimestampNow(),
-						Event: &bepb.BuildEvent_ComponentStreamFinished{ComponentStreamFinished: &bepb.BuildEvent_BuildComponentStreamFinished{
-							Type: bepb.BuildEvent_BuildComponentStreamFinished_FINISHED,
-						}},
-					},
+					SequenceNumber: i + 1,
+					Event:          event,
 				},
-			})
-			log.Debugf("BEP: published FINISHED event (#%d) in %s", seqNo, time.Since(start))
-
-			if err != nil {
-				p.setError(err)
-				return
 			}
+			i++
+			if err := stream.Send(req); err != nil {
+				return status.UnavailableErrorf("send failed: %s", err)
+			}
+			// After successfully transmitting all events, close our side of the stream
+			// and wait for server ACKs before closing the connection.
+			if _, ok := event.Event.(*bepb.BuildEvent_ComponentStreamFinished); ok {
+				if err := stream.CloseSend(); err != nil {
+					return status.UnavailableErrorf("close send failed: %s", err)
+				}
+				finished = true
+			}
+		}
+		if finished {
 			break
 		}
+		// Wait until another event has been added to the buffer since we called
+		// Get(), then rinse and repeat.
+		p.events.Wait()
+	}
+
+	// Return any error from receiving ACKs
+	return <-doneReceiving
+}
 
+func (p *Publisher) Publish(event *bespb.BuildEvent) error {
+	be := &bepb.BuildEvent{EventTime: ptypes.TimestampNow()}
+	if event == nil {
+		be.Event = &bepb.BuildEvent_ComponentStreamFinished{ComponentStreamFinished: &bepb.BuildEvent_BuildComponentStreamFinished{
+			Type: bepb.BuildEvent_BuildComponentStreamFinished_FINISHED,
+		}}
+	} else {
 		bazelEvent, err := ptypes.MarshalAny(event)
 		if err != nil {
-			p.setError(status.WrapError(err, "failed to marshal bazel event"))
-			return
-		}
-		start := time.Now()
-		err = stream.Send(&pepb.PublishBuildToolEventStreamRequest{
-			OrderedBuildEvent: &pepb.OrderedBuildEvent{
-				StreamId:       p.streamID,
-				SequenceNumber: seqNo,
-				Event: &bepb.BuildEvent{
-					EventTime: ptypes.TimestampNow(),
-					Event:     &bepb.BuildEvent_BazelEvent{BazelEvent: bazelEvent},
-				},
-			},
-		})
-		log.Debugf("BEP: published event (#%d) in %s", seqNo, time.Since(start))
-		if err != nil {
-			p.setError(err)
-			return
+			return err
 		}
+		be.Event = &bepb.BuildEvent_BazelEvent{BazelEvent: bazelEvent}
 	}
-	// After successfully transmitting all events, close our side of the stream
-	// and wait for server ACKs before closing the connection.
-	if err := stream.CloseSend(); err != nil {
-		p.setError(status.WrapError(err, "failed to close build event stream"))
-		return
-	}
-	<-doneReceiving
+
+	p.events.Add(be)
+	return nil
 }
 
-func (p *Publisher) Publish(e *bespb.BuildEvent) error {
-	p.mu.Lock()
-	defer p.mu.Unlock()
-	if p.err != nil {
-		return status.WrapError(p.err, "cannot publish event due to previous error")
+// Wait publishes the Finished event and then waits for the server to ACK
+// all published events.
+func (p *Publisher) Wait() error {
+	if err := p.Publish(nil); err != nil {
```

Comment:
Instead of treating `nil` as a special value, why don't we have this function create the finished event?